lntemp/node/watcher.go

```diff
@@ -0,0 +1,679 @@
+	// openChanWatchers  map[wire.OutPoint][]chan struct{}
+	openChanWatchers *sync.Map
+
+	// closeChanWatchers map[wire.OutPoint][]chan struct{}
+	closeChanWatchers *sync.Map
+
+	wg sync.WaitGroup
+}
+
+func newNodeWatcher(rpc *rpc.HarnessRPC, state *State) *nodeWatcher {
+	return &nodeWatcher{
+		rpc:               rpc,
+		state:             state,
+		chanWatchRequests: make(chan *chanWatchRequest, 100),
+		openChanWatchers:  &sync.Map{},
+		closeChanWatchers: &sync.Map{},
+	}
+}
+
+// GetNumChannelUpdates reads the num of channel updates inside a lock and
+// returns the value.
+func (nw *nodeWatcher) GetNumChannelUpdates(op wire.OutPoint) int {
+	result, ok := nw.state.numChanUpdates.Load(op)
+	if ok {
+		return result.(int)
+	}
+	return 0
+}
+
+// GetPolicyUpdates returns the node's policyUpdates state.
+func (nw *nodeWatcher) GetPolicyUpdates(op wire.OutPoint) PolicyUpdate {
+	result, ok := nw.state.policyUpdates.Load(op)
+	if ok {
+		return result.(PolicyUpdate)
+	}
+	return nil
+}
+
+// GetNodeUpdates reads the node updates inside a lock and returns the value.
+func (nw *nodeWatcher) GetNodeUpdates(pubkey string) []*lnrpc.NodeUpdate {
+	result, ok := nw.state.nodeUpdates.Load(pubkey)
+	if ok {
+		return result.([]*lnrpc.NodeUpdate)
+	}
+	return nil
+}
+
+// WaitForNumChannelUpdates will block until a given number of updates has been
+// seen in the node's network topology.
+func (nw *nodeWatcher) WaitForNumChannelUpdates(op wire.OutPoint,
+	expected int) error {
+
+	ticker := time.NewTicker(wait.PollInterval)
+	timer := time.After(DefaultTimeout)
+	defer ticker.Stop()
+
+	for {
+		num := nw.GetNumChannelUpdates(op)
+
+		select {
+		// Check the num every 200ms.
+		case <-ticker.C:
+			if num >= expected {
+				return nil
+			}
+
+		case <-timer:
+			return fmt.Errorf("timeout waiting for num channel "+
+				"updates, want %d, got %d", expected, num)
+		}
+	}
+}
+
+// WaitForNumNodeUpdates will block until a given number of node updates has
+// been seen in the node's network topology.
+func (nw *nodeWatcher) WaitForNumNodeUpdates(pubkey string,
+	expected int) error {
+
+	ticker := time.NewTicker(wait.PollInterval)
+	timer := time.After(DefaultTimeout)
+	defer ticker.Stop()
+
+	for {
+		num := len(nw.GetNodeUpdates(pubkey))
+
+		select {
+		// Check the num every 200ms.
+		case <-ticker.C:
+			if num >= expected {
+				return nil
+			}
+
+		case <-timer:
+			return fmt.Errorf("timeout waiting for num node "+
+				"updates, want %d, got %d", expected, num)
+		}
+	}
+}
+
+// WaitForChannelOpen will block until a channel with the target outpoint is
+// seen as being fully advertised within the network. A channel is considered
+// "fully advertised" once both of its directional edges has been advertised in
+// the node's network topology.
+func (nw *nodeWatcher) WaitForChannelOpen(chanPoint *lnrpc.ChannelPoint) error {
+	op := nw.rpc.MakeOutpoint(chanPoint)
+	eventChan := make(chan struct{})
+	nw.chanWatchRequests <- &chanWatchRequest{
```

Comment:
Do we need to select on a quit/timeout here so we don't risk hanging if a test fails unexpectedly?

Reply:
yep and we also buffers it during initialization.